File: Python/Stack/636. Exclusive Time of Functions.py
# The problem description is misleading, basically, the following schedule is not allowed:

# ----111111
# 000000----

# It must be like this:

# ---1111---
# --000000--


class Solution:
    
    def exclusiveTime(self, N, logs):
        ans = [0] * N
        stack = []
        prev_time = 0

        for log in logs:
            fn, typ, time = log.split(':')
            fn, time = int(fn), int(time)

            if typ == 'start':
                if stack:
                    ans[stack[-1]] += time - prev_time 
                stack.append(fn)
                prev_time = time
            else:
                ans[stack.pop()] += time - prev_time + 1
                prev_time = time + 1

        return ans

File: C++/Stack/636. Exclusive Time of Functions.cpp
class Solution {
public:
    
    vector<string> split(string& s, string del = " ") {
        vector<string> res;
        int start = 0;
        int end = s.find(del);
        while (end != -1) {
            res.push_back(s.substr(start, end - start));
            start = end + del.size();
            end = s.find(del, start);
        }
        res.push_back(s.substr(start, end - start));
        return res;
    }
    
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        stack<int> stk;
        vector<int> ans(n, 0);
        int time = 0;
        for (string& log : logs) {
            vector<string> s = split(log, ":");
            int id = stoi(s[0]);
            string fn = s[1];
            int timestamp = stoi(s[2]);
            if (fn == "start") {
                if (!stk.empty()) ans[stk.top()] += timestamp - time;
                stk.push(id);
                time = timestamp;
            }
            else {
                ans[stk.top()] += timestamp - time + 1;
                stk.pop();
                time = timestamp + 1;
            }
        }
        return ans;
    }
};